app/models/user.rb

```diff
@@ -25,23 +29,47 @@ def self.from_omniauth(auth)
     end
   end
 
-  def friends
-    User::where(id: senders.friend.pluck(:receiver_id) + receivers.friend.pluck(:sender_id))
+  def joined_groups
+    Group.where(id: self.member_relations.joined.pluck(:user_id))
   end
 
-  def home_posts
-    Post::where(user_id: self.id).or(Post::where(user_id: self.friends.pluck(:id)))
+  def friends
+    User.where('id in (?) or id in (?)', self.senders.friend.pluck(:receiver_id), self.receivers.friend.pluck(:sender_id))
   end
 
-  def relations
-    Relation::where('sender_id=? or receiver_id=?', self.id, self.id)
+  def home_posts
+    Post.where('user_id=? or user_id in (?)', self.id, self.friends.pluck(:id)).order(created_at: :desc)
   end
 
   def relation(check_user_id)
-    Relation::where('sender_id=? or receiver_id=?', self.id, self.id).where('sender_id=? or receiver_id=?', check_user_id, check_user_id).first
+    Relation::where('sender_id in (?) and receiver_id in (?)', [self.id, check_user_id], [self.id, check_user_id]).first
   end
 
   def common_friends(check_user_id)
-    User.where(id: self.friends.pluck(:id)).where(id: User.find(check_user_id).friends.pluck(:id))
+    User.where('id in (?) and id in (?)', self.friends.pluck(:id), User.find(check_user_id).friends.pluck(:id))
```

Comment:
ko sai nhưng viết khó hiểu, đôi khi có thể viết tách ra để người đọc có thể hiểu đc mình đang làm những gì        
```ruby
friend_ids = self.friends.pluck(:id)
check_user_friend_ids = User.find(check_user_id).friends.pluck(:id)  
common_friend_ids = (friend_ids + check_user_friend_ids).uniq 
User.where(id: common_friend_ids) # cũng có thể gộp dòng 3 với 4 làm một
```

Reply:
em thấy có 2 logic đều gọi ra friend của 1 user nên em gộp lại, em sẽ mở rộng cái hàm friend của 1 user ra nhiều user

def common_friends(check_user_id)
    User.where(id: Relation.by_sender([self.id, check_user_id]).friend.pluck(:sender_id))
      .or(User.where(id: Relation.by_receiver([self.id, check_user_id]).friend.pluck(:receiver_id)))
      .where.not(id: [self.id, check_user_id])
      .order(created_at: :desc)
  end